rust-runtime/aws-smithy-http/src/event_stream/sender.rs

```diff
@@ -4,20 +4,25 @@
  */
 
 use crate::result::SdkError;
+use aws_smithy_async::future::fn_stream::FnStream;
 use aws_smithy_eventstream::frame::{MarshallMessage, SignMessage};
 use bytes::Bytes;
-use futures_core::Stream;
 use std::error::Error as StdError;
 use std::fmt;
 use std::fmt::Debug;
+use std::future::poll_fn;
 use std::marker::PhantomData;
 use std::pin::Pin;
+use std::sync::Mutex;
 use std::task::{Context, Poll};
 use tracing::trace;
 
 /// Input type for Event Streams.
 pub struct EventStreamSender<T, E> {
-    input_stream: Pin<Box<dyn Stream<Item = Result<T, E>> + Send + Sync>>,
+    // `FnStream` does not have a `Sync` bound but this struct needs to be `Sync`
+    // as demonstrated by a unit test `event_stream_sender_send_sync`.
+    // Wrapping `input_stream` with a `Mutex` will make `EventStreamSender` `Sync`.
+    input_stream: Mutex<FnStream<Result<T, E>>>,
```

Comment:
Was there a reason that the boxed future inside FnStream doesn't have a Sync bound? Is it just to make implementing the stream easier?

Reply:
Yes, the underlying `generator` is a future that is only `Send`, which is usually the case for `Future` ([discussion](https://users.rust-lang.org/t/future-is-not-sync-when-calling-async-fn-of-dyn-trait-obj-in-another-async-fn/83285/2)). Added a comment in d2e4411.

Reply:
Why does `EventStreamSender` need to be `Sync`? i.e. why does that test exist?